resources/net.go

```diff
@@ -0,0 +1,726 @@
+		return err // bubble up a NACK...
+	}
+
+	var send = false
+	var exit *error
+
+	for {
+		select {
+		case event, ok := <-ifaceChan:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			// save the operational state to the struct
+			states, ok := event.Body[0].(map[string]dbus.Variant)
+			if !ok {
+				return fmt.Errorf("error casting states")
+			}
+			if val, ok := states[OpState]; ok {
+				obj.OpState = val
+			}
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event.Sender, event.Body)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event, ok := <-recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s watcher error", obj)
+			}
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event, ok := <-svcChan:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event[NetworkdService].ActiveState, event)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case err, ok := <-svcErr:
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			return errwrap.Wrapf(err, "unknown %s error", obj)
+
+		case event, ok := <-obj.Events():
+			if !ok {
+				return fmt.Errorf("channel closed unexpectedly")
+			}
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// ifaceCheckApply checks the state of the network device and performs the
+// appropriate actions to converge its state with the graph.
+func (obj *NetRes) ifaceCheckApply(apply bool) (checkOK bool, _ error) {
+	var ok bool
+
+	// create a bus object pointing to the adapter
+	busObj := obj.gdbus.Object(DBusNetworkdInterface, obj.IfacePath)
+	// get the object's operational state
+	call := busObj.Call(DBusPropertiesGet, 0, DBusNetworkdLinkInterface, OpState)
+	if call.Err != nil {
+		return false, call.Err
+	}
+	// save the operational state to the struct
+	obj.OpState, ok = call.Body[0].(dbus.Variant)
+	if !ok {
+		return false, fmt.Errorf("error casting opstate")
+	}
+
+	// if the interface's state is correct, we're done
+	if obj.State == IfaceDown && obj.OpState == dbus.MakeVariant(OpStateOff) {
+		return true, nil
+	}
+	if obj.State == IfaceUp {
+		if obj.OpState == dbus.MakeVariant(OpStateRoutable) {
+			return true, nil
+		}
+		// If the connection is lost and timeout is set, retry once a second
+		// until the timeout expires. If the connection comes back, continue
+		// as if nothing happened. Error if the timeout expires (or isn't set.)
+		if obj.OpState == dbus.MakeVariant(OpStateNoCarrier) {
+			for i := 0; i < int(obj.Timeout); i++ {
```

Comment:
I need some mentoring on making this exit cleanly. I'll try to figure it out in the meantime.

Reply:
Happy to-- let's discuss why we need a timeout at all... I think explaining this in the struct will be the best starting point, although yes, I agree having a blocking `time.Sleep` in the code is not ideal.